js/runner_test.go

```diff
@@ -268,6 +268,161 @@ func TestSetupTeardown(t *testing.T) {
 	}
 }
 
+func TestSetupDataIsolation(t *testing.T) {
+	r1, err := New(&lib.SourceData{
+		Filename: "/script.js",
+		Data: []byte(`
+			export let options = {
+				setupTimeout: "10s",
+				teardownTimeout: "10s",
+			};
+
+			export function setup() {
+				return { v: 1 };
+			}
+			export function teardown(data) {
+				if (data.v != 1) {
+					throw new Error("teardown: wrong data: " + data.v)
+				}
+				data.v = 2
+			}
+			export default function(data) {
+				if (data.v != 1) {
+					throw new Error("default: wrong data: " + JSON.stringify(data))
+				}
+				data.v = 2
+			}
+		`),
+	}, afero.NewMemMapFs(), lib.RuntimeOptions{})
+	if !assert.NoError(t, err) {
+		return
+	}
+
+	r2, err := NewFromArchive(r1.MakeArchive(), lib.RuntimeOptions{})
+	if !assert.NoError(t, err) {
+		return
+	}
+
+	testdata := map[string]*Runner{"Source": r1, "Archive": r2}
+	for name, r := range testdata {
+		samples := make(chan stats.SampleContainer, 100)
+		t.Run(name, func(t *testing.T) {
+			if !assert.NoError(t, r.Setup(context.Background(), samples)) {
+				return
+			}
+
+			vu, err := r.NewVU(samples)
+			if assert.NoError(t, err) {
+				for i := 0; i < 10; i++ {
+					err := vu.RunOnce(context.Background())
+					assert.NoError(t, err)
+				}
+			}
+
+			assert.NoError(t, r.Teardown(context.Background(), samples))
+		})
+	}
+}
+
+func testSetupDataHelper(t *testing.T, src *lib.SourceData) {
+	expScriptOptions := lib.Options{SetupTimeout: types.NullDurationFrom(1 * time.Second)}
+	r1, err := New(src, afero.NewMemMapFs(), lib.RuntimeOptions{Env: map[string]string{"expectedSetupTimeout": "1s"}})
+	require.NoError(t, err)
+	require.Equal(t, expScriptOptions, r1.GetOptions())
+
+	r2, err := NewFromArchive(r1.MakeArchive(), lib.RuntimeOptions{Env: map[string]string{"expectedSetupTimeout": "3s"}})
+	require.NoError(t, err)
+	require.Equal(t, expScriptOptions, r2.GetOptions())
+
+	newOptions := lib.Options{SetupTimeout: types.NullDurationFrom(3 * time.Second)}
+	r2.SetOptions(newOptions)
+	require.Equal(t, newOptions, r2.GetOptions())
+
+	testdata := map[string]*Runner{"Source": r1, "Archive": r2}
+	for name, r := range testdata {
+		t.Run(name, func(t *testing.T) {
+			samples := make(chan stats.SampleContainer, 100)
+
+			if !assert.NoError(t, r.Setup(context.Background(), samples)) {
+				return
+			}
+			vu, err := r.NewVU(samples)
+			if assert.NoError(t, err) {
+				err := vu.RunOnce(context.Background())
+				assert.NoError(t, err)
+			}
+		})
+	}
+}
+func TestSetupDataReturnValue(t *testing.T) {
```

Comment:
I think this pretty much duplicates `TestSetupTeardown()` in the same file

Reply:
I didn't want to have a test for two things, but if setupData will be shared between iterations of the same VU, This test should problably run multiple VUs in parallel instead

Reply:
Hmm yeah, good point, but this test doesn't actually modify `data` at all, it just checks if it's 42. `TestSetupDataIsolation()` on the other hand would have to be modified. And yeah, since you'd start multiple VUs concurrently, they might as well modify the same `setupData` element concurrently, to fail when there are data races as well. Obviously with the current implementation that won't happen, but just in case, so we don't backslide in the future and forget what all of that json unmarshalling is for :smile: